File: Languages/cpp/algorithms-dsa/leetcode/bit-manipulation/number_of_1_bits.cpp
/**
 * LeetCode 191: Number of 1 Bits
 * Topic: Bit Manipulation
 * Difficulty: Easy
 *
 * Return the number of set bits (1-bits) in an unsigned integer.
 * Use n &= (n-1) trick to clear the lowest set bit each iteration.
 * Time: O(k) where k = number of set bits, Space: O(1)
 */
#include <iostream>
#include <cstdint>
#include <cassert>
using namespace std;

int hammingWeight(uint32_t n) {
    int count = 0;
    while (n) {
        n &= (n - 1);
        count++;
    }
    return count;
}

int main() {
    assert(hammingWeight(0b00000000000000000000000000001011) == 3);
    assert(hammingWeight(0b00000000000000000000000010000000) == 1);
    assert(hammingWeight(0b11111111111111111111111111111101) == 31);
    assert(hammingWeight(0) == 0);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/bit-manipulation/number_of_1_bits.py
"""
LeetCode #191 - Number of 1 Bits
Topic: Bit Manipulation
Difficulty: Easy

Count set bits using n & (n-1) trick to clear lowest set bit.

Time Complexity: O(k) where k = number of set bits
Space Complexity: O(1)
"""


def hamming_weight(n: int) -> int:
    count = 0
    while n:
        n &= n - 1
        count += 1
    return count


if __name__ == "__main__":
    assert hamming_weight(0b00000000000000000000000000001011) == 3
    assert hamming_weight(0b00000000000000000000000010000000) == 1
    assert hamming_weight(0) == 0
    assert hamming_weight(0b11111111111111111111111111111101) == 31
    print("All tests passed!")
